pkg/querier/batch/stream_test.go

```diff
@@ -180,3 +181,358 @@ func requireBatchEqual(t *testing.T, b, o chunk.Batch) {
+	t.Run("same timestamp different float values", func(t *testing.T) {
+		s := newBatchStream(2, nil, nil)
+
+		// Create batch with value 10.0 at timestamp 100
+		leftBatch := mkGenericFloatBatch(100, 1)
+		leftBatch.Values[0] = 10.0
+		s.batches = []chunk.Batch{leftBatch}
+
+		// Create batch with value 5.0 at same timestamp 100
+		rightBatch := mkGenericFloatBatch(100, 1)
+		rightBatch.Values[0] = 5.0
+
+		s.merge(&rightBatch, chunk.BatchSize, 0)
+
+		require.Equal(t, 1, len(s.batches))
+		require.Equal(t, 1, s.batches[0].Length)
+		require.Equal(t, int64(100), s.batches[0].Timestamps[0])
+		require.Equal(t, 10.0, s.batches[0].Values[0]) // Higher value should win
+	})
+
+	t.Run("multiple timestamp conflicts", func(t *testing.T) {
+		s := newBatchStream(3, nil, nil)
+
+		// Create batch with values at timestamps 100, 200, 300
+		leftBatch := mkGenericFloatBatch(100, 3)
+		leftBatch.Values[0] = 10.0 // ts 100
+		leftBatch.Values[1] = 20.0 // ts 101
+		leftBatch.Values[2] = 30.0 // ts 102
+		s.batches = []chunk.Batch{leftBatch}
+
+		// Create batch with conflicting values at same timestamps
+		rightBatch := mkGenericFloatBatch(100, 3)
+		rightBatch.Values[0] = 15.0 // ts 100 - higher, should win
+		rightBatch.Values[1] = 15.0 // ts 101 - lower, should lose
+		rightBatch.Values[2] = 25.0 // ts 102 - lower, should lose
+
+		s.merge(&rightBatch, chunk.BatchSize, 0)
+
+		require.Equal(t, 1, len(s.batches))
+		require.Equal(t, 3, s.batches[0].Length)
+		require.Equal(t, 15.0, s.batches[0].Values[0]) // Right wins (15.0 > 10.0)
+		require.Equal(t, 20.0, s.batches[0].Values[1]) // Left wins (20.0 > 15.0)
+		require.Equal(t, 30.0, s.batches[0].Values[2]) // Left wins (30.0 > 25.0)
+	})
+}
+
+func mkGenericFloatHistogramBatch(from int64, size int) chunk.Batch {
+	batch := chunk.Batch{ValueType: chunkenc.ValFloatHistogram}
+	for i := 0; i < size; i++ {
+		batch.Timestamps[i] = from + int64(i)
+		h := test.GenerateTestFloatHistogram(int(from) + i)
+		batch.PointerValues[i] = unsafe.Pointer(h)
+	}
+	batch.Length = size
+	return batch
+}
+
+func TestBatchStream_EdgeCases(t *testing.T) {
+	t.Run("empty batch merge", func(t *testing.T) {
+		s := newBatchStream(2, nil, nil)
+
+		// Start with empty stream
+		require.Equal(t, 0, s.len())
+
+		// Merge with empty batch
+		emptyBatch := chunk.Batch{ValueType: chunkenc.ValFloat, Length: 0}
+		s.merge(&emptyBatch, chunk.BatchSize, 0)
+		// Empty batches might still get added, but they should not contain data
+		if s.len() > 0 {
+			require.Equal(t, 0, s.batches[0].Length)
+		}
+
+		// Merge non-empty batch with empty stream
+		batch := mkGenericFloatBatch(100, 2)
+		s.merge(&batch, chunk.BatchSize, 0)
+		require.Equal(t, 1, s.len())
+		require.Equal(t, 2, s.batches[0].Length)
+
+		// Merge empty batch with non-empty stream
+		s.merge(&emptyBatch, chunk.BatchSize, 0)
+		require.Equal(t, 1, s.len())
+		require.Equal(t, 2, s.batches[0].Length)
+	})
+
+	t.Run("single sample batches with conflicts", func(t *testing.T) {
```

Comment:
[nit] This looks to be the same test case as `same timestamp different float values` in `TestBatchStream_MergeWithTimestampConflicts`. I would remove this.